

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Gallery of examples of MacSyFinder’s models &mdash; MacSyFinder 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Carrying models from v1 to v2" href="models_V1ToV2.html" />
    <link rel="prev" title="Publishing/sharing models" href="publish_package.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MacSyFinder
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/index.html">User Guide</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Modeller Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#modelling-systems-with-macsyfinder">Modelling Systems with MacSyFinder</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="package.html">Models Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="modeling.html">Macromolecular models</a></li>
<li class="toctree-l3"><a class="reference internal" href="helper_tools.html">Helper Tool</a></li>
<li class="toctree-l3"><a class="reference internal" href="publish_package.html">Publishing/sharing models</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Gallery of examples of MacSyFinder’s models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#getting-started-with-a-not-so-simple-example-modelling-the-t1ss">Getting started with a (not-so-)simple example: modelling the T1SS</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-case-of-t3ss-and-bacterial-flagella-or-how-to-distinguish-homologous-cellular-machineries">The case of T3SS and bacterial flagella, or how to distinguish homologous cellular machineries</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#carrying-models-from-v1-to-v2">Carrying models from v1 to v2</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Modeller Guide</a> &raquo;</li>
        
      <li>Gallery of examples of MacSyFinder’s models</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/modeler_guide/gallery.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gallery-of-examples-of-macsyfinder-s-models">
<span id="gallery-models"></span><h1>Gallery of examples of MacSyFinder’s models<a class="headerlink" href="#gallery-of-examples-of-macsyfinder-s-models" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="table-of-contents-of-gallery">
<p class="topic-title first">Table of contents of gallery</p>
<ul class="simple">
<li><p><a class="reference internal" href="#getting-started-with-a-not-so-simple-example-modelling-the-t1ss" id="id9">Getting started with a (not-so-)simple example: modelling the T1SS</a></p></li>
<li><p><a class="reference internal" href="#the-case-of-t3ss-and-bacterial-flagella-or-how-to-distinguish-homologous-cellular-machineries" id="id10">The case of T3SS and bacterial flagella, or how to distinguish homologous cellular machineries</a></p></li>
</ul>
</div>
<p>Here follows a “gallery” of MacSyFinder models we have developed over the years, attempting to describe the reasoning behind the modeling process.</p>
<p>These examples are extracted from published work, see the following references (they include more examples):</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.nature.com/articles/srep23080">Abby et al. 2016</a>, <em>Scientific Reports</em>, for the description of the T1SS and T3SS models (and way more models not discussed here).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a>, <em>PLoS Genetics</em>, for the evolutionary study of the T3SS and the bacterial flagellum, and how were designed the corresponding profiles.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pbio.3000390">Denise et al. 2019</a>, <em>PLoS Biology</em>, for the description of the T2SS and type IV-filament super-family models.</p></li>
</ul>
<div class="section" id="getting-started-with-a-not-so-simple-example-modelling-the-t1ss">
<span id="t1ss"></span><h2><a class="toc-backref" href="#id9">Getting started with a (not-so-)simple example: modelling the T1SS</a><a class="headerlink" href="#getting-started-with-a-not-so-simple-example-modelling-the-t1ss" title="Permalink to this headline">¶</a></h2>
<div class="section" id="identifying-genetic-components">
<h3>1. Identifying genetic components<a class="headerlink" href="#identifying-genetic-components" title="Permalink to this headline">¶</a></h3>
<p>The type I secretion system (T1SS) consists in three conserved components:</p>
<ul class="simple">
<li><p>an ABC transporter (ABC)</p></li>
<li><p>a membrane-fusion protein (MFP)</p></li>
<li><p>an outer membrane protein (OMF)</p></li>
</ul>
<p>For their detection, we therefore need to provide HMM profiles for each component, for example: “abc.hmm”, “mfp.hmm” and “omf.hmm”.
These can be specifically designed, or taken from HMM profiles databanks such as <a class="reference external" href="http://pfam.xfam.org/">PFAM</a> , <a class="reference external" href="http://tigrfams.jcvi.org/cgi-bin/index.cgi">TIGRFAM</a> or <a class="reference external" href="https://supfam.mrc-lmb.cam.ac.uk/SUPERFAMILY/">SUPERFAMILY</a>..</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For suggestions on how to design specific HMM protein profiles, read our dedicated book chapter:</p>
<blockquote>
<div><p><a class="reference external" href="https://link.springer.com/protocol/10.1007/978-1-4939-7033-9_1">Identification of Protein Secretion Systems in Bacterial Genomes Using MacSyFinder</a> by Sophie Abby and Eduardo Rocha, in <em>Methods in Molecular Biology</em> (2017).</p>
</div></blockquote>
</div>
</div>
<div class="section" id="determining-the-role-of-the-components">
<h3>2. Determining the role of the components<a class="headerlink" href="#determining-the-role-of-the-components" title="Permalink to this headline">¶</a></h3>
<p>From litterature, the three components listed above <em>must</em> be present to have a viable T1SS. Therefore, these are all deemed <em>mandatory</em> in the model of the T1SS.</p>
</div>
<div class="section" id="describing-their-genetic-architecture">
<h3>3. Describing their genetic architecture<a class="headerlink" href="#describing-their-genetic-architecture" title="Permalink to this headline">¶</a></h3>
<p>According to the litteraure, the genes encoding the three components listed above are generally found lying next to each other in genomes. Therefore, these are considered as “single-locus” system. In addition, there is the particular case of the OMF component. It can either be found:</p>
<ul class="simple">
<li><p>next to the two other components, as explained just below</p></li>
<li><p>in some other cases, it can be involved in other cellular machineries functioning, and thus be encoded some place else that at the main T1SS’ locus (in this case, made of ABC+MFP).</p></li>
</ul>
<p>Therefore, we can attribute the <cite>loner</cite> feature to the OMF component.</p>
<p>In addition to the latter exception described, it means that this OMF component can also be involved in the functioning of not a single, but several machineries at the same time. In practice, this would mean that two full sets of T1SS components can be inferred with a single OMF component found in the genome. This corresponds to the <cite>multi-system</cite> feature.</p>
<a class="reference internal image-reference" href="../_images/T1SS_example.svg"><img alt="../_images/T1SS_example.svg" class="align-center" height="4000px" src="../_images/T1SS_example.svg" /></a>
</div>
<div class="section" id="writing-down-the-model">
<h3>4. Writing down the model<a class="headerlink" href="#writing-down-the-model" title="Permalink to this headline">¶</a></h3>
<p>Now that all elements of the model are listed, the model for the T1SS can be written using the dedicated MacSyFinder XML grammar:</p>
<div class="highlight-xml notranslate"><div class="highlight"><pre><span></span><span class="nt">&lt;model</span> <span class="na">inter_gene_max_space=</span><span class="s">&quot;5&quot;</span> <span class="na">min_mandatory_genes_required=</span><span class="s">&quot;3&quot;</span> <span class="na">min_genes_required=</span><span class="s">&quot;3&quot;</span> <span class="na">vers=</span><span class="s">&quot;2.0&quot;</span><span class="nt">&gt;</span>
    <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T1SS_abc&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
    <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T1SS_mfp&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
    <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T1SS_omf&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span> <span class="na">loner=</span><span class="s">&quot;1&quot;</span> <span class="na">multi_system=</span><span class="s">&quot;1&quot;</span><span class="nt">/&gt;</span>
<span class="nt">&lt;/model&gt;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="the-case-of-t3ss-and-bacterial-flagella-or-how-to-distinguish-homologous-cellular-machineries">
<span id="t3ss"></span><h2><a class="toc-backref" href="#id10">The case of T3SS and bacterial flagella, or how to distinguish homologous cellular machineries</a><a class="headerlink" href="#the-case-of-t3ss-and-bacterial-flagella-or-how-to-distinguish-homologous-cellular-machineries" title="Permalink to this headline">¶</a></h2>
<p>The type III secretion system (T3SS), involved in proteic effectors secretion into eukaryotic cells) and the bacterial flagellum (involved in motility) are evolutionarily related (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a>). This can make their annotation in genomes tricky, if only based on core components that can have homologs in both systems.</p>
<p>However, these machineries also have <strong>specific core components</strong>. With MacSyFinder and the <em>forbidden</em> feature for components, it is possible to model this, and create models for efficient discrimination between homologous machineries.</p>
<p>For a toy example on how to model similar yet distinct machineries, you can also have a look <a class="reference internal" href="modeling.html#model-definition-grammar-label"><span class="std std-ref">here</span></a>.</p>
<div class="section" id="identifying-genetic-components-and-determining-their-role">
<h3>1. Identifying genetic components and determining their role<a class="headerlink" href="#identifying-genetic-components-and-determining-their-role" title="Permalink to this headline">¶</a></h3>
<p>The T3SS is partly homologous to the bacterial flagellum: 8 of its 9 core components are homologous to core components of the flagellum. This is explained by the fact that the T3SS is evolutionarily derived from the flagellum (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a>).
Yet, the T3SS is made of two dozens of components, and the flagellum, more than twice this number of components:</p>
<ul class="simple">
<li><p>The flagellum presents specific core components that have no counterpart in the T3SS.</p></li>
<li><p>It is also the case of the T3SS, which has one specific core component: the secretin.</p></li>
</ul>
<p>Solely based on the specificity of core components, it is possible to distinguish T3SS from flagella. This can be done by listing the <strong>specific core components</strong> of a given system as <em>mandatory</em> in the system, and as <em>forbidden</em> in the homologous system.</p>
<p>Then, HMM protein profiles can be specifically designed for these components, or can be retrieved from databases such as <a class="reference external" href="http://pfam.xfam.org/">PFAM</a> , <a class="reference external" href="http://tigrfams.jcvi.org/cgi-bin/index.cgi">TIGRFAM</a> or <a class="reference external" href="https://supfam.mrc-lmb.cam.ac.uk/SUPERFAMILY/">SUPERFAMILY</a>.</p>
</div>
<div class="section" id="dealing-with-components-with-varied-evolutionary-origins">
<h3>2. Dealing with components with varied evolutionary origins<a class="headerlink" href="#dealing-with-components-with-varied-evolutionary-origins" title="Permalink to this headline">¶</a></h3>
<p>Another peculiarity of T3SS’ evolutionary history consists in that of the secretin, which has been co-opted (acquired) at least three times independently along T3SS diversification: once from the T2SS, once from the Tad pilus, and once from the Type IVa pilus (<a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a> , <a class="reference external" href="https://doi.org/10.1371/journal.pbio.3000390">Denise et al. 2019</a> ).</p>
<p>This means that sometimes, the T3SS secretin will have more sequence similarity for the secretins from these other machineries - and thus that the profile for the T3SS secretin might “miss” these components, whereas profiles for secretins from the T2SS, T4P or Tad might be more efficient to retrieve them.</p>
<p>Using the <em>exchangeables</em> feature, MacSyFinder enables to use different HMM protein profile to search for components that may fill a same function. Therefore, it is possible to list profiles of secretins from other machineries among the set of profiles to use to retrieve all T3SS potential secretins.</p>
<p>In the following drawing, a scheme of a T3SS is shown on the left, and the features listed above are shown on a scheme of the T3SS model, including forbidden components from the flagellum (red crosses), and exchangeable components for the secretin “sctC”, depicted with yellow boxes (with the name of the secretin gene from the T4aP, T2SS and Tad pilus respectively). The <em>inter-gene-max-space</em> parameter - i.e., maximal number of components allowed between two systems’ components to consider them consecutive - is expressed with the “d” letter.</p>
<a class="reference internal image-reference" href="../_images/T3SS_example.svg"><img alt="../_images/T3SS_example.svg" class="align-center" height="4000px" src="../_images/T3SS_example.svg" /></a>
</div>
<div class="section" id="describing-the-quorum-and-genetic-architecture-of-the-systems">
<h3>3. Describing the quorum, and genetic architecture of the systems<a class="headerlink" href="#describing-the-quorum-and-genetic-architecture-of-the-systems" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>T3SS and bacterial flagella are generally encoded on the form of multi-components loci in genomes. Given the fact that we designed HMM protein profiles only for the most conserved, core components of these machineries, and that it means that several systems’ components can intersperse between the core ones (remember, T3SS has aroound 25 components, and the flagellum &gt;40), we set the <em>inter-gene-max-space</em> parameter (maximal number of components allowed between two systems’ components to consider them consecutive) to 10 in the case of the T3SS, and to 20 in the case of the flagellum.</p></li>
<li><p>T3SS and bacterial flagella can be encoded by one, or multiple loci. We therefore use the <em>multi-loci</em> feature to describe their genetic architecture (set to “1”, meaning “True” in the models).</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For suggestions on how to set the quorum and genetic architecture parameters, read our dedicated book chapter:</p>
<blockquote>
<div><p><a class="reference external" href="https://link.springer.com/protocol/10.1007/978-1-4939-7033-9_1">Identification of Protein Secretion Systems in Bacterial Genomes Using MacSyFinder</a> by Sophie Abby and Eduardo Rocha, in <em>Methods in Molecular Biology</em> (2017).</p>
</div></blockquote>
</div>
</div>
<div class="section" id="writing-down-the-models">
<h3>4. Writing down the models<a class="headerlink" href="#writing-down-the-models" title="Permalink to this headline">¶</a></h3>
<p>Given all the features described above, here is the model of the T3SS:</p>
<p><strong>T3SS.xml</strong></p>
<div class="highlight-xml notranslate"><div class="highlight"><pre><span></span><span class="nt">&lt;model</span> <span class="na">inter_gene_max_space=</span><span class="s">&quot;10&quot;</span> <span class="na">min_mandatory_genes_required=</span><span class="s">&quot;7&quot;</span> <span class="na">min_genes_required=</span><span class="s">&quot;7&quot;</span> <span class="na">multi_loci=</span><span class="s">&quot;1&quot;</span> <span class="na">vers=</span><span class="s">&quot;2.0&quot;</span><span class="nt">&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctC&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">&gt;</span>
       <span class="nt">&lt;exchangeables&gt;</span>
          <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T2SS_gspD&quot;</span><span class="nt">/&gt;</span>
          <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T4P_pilQ&quot;</span><span class="nt">/&gt;</span>
          <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Tad_rcpA&quot;</span><span class="nt">/&gt;</span>
       <span class="nt">&lt;/exchangeables&gt;</span>
   <span class="nt">&lt;/gene&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctJ&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctN&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctQ&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctR&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctS&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctT&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctU&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctV&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_fliE&quot;</span> <span class="na">presence=</span><span class="s">&quot;forbidden&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_flgB&quot;</span> <span class="na">presence=</span><span class="s">&quot;forbidden&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_flgC&quot;</span> <span class="na">presence=</span><span class="s">&quot;forbidden&quot;</span><span class="nt">/&gt;</span>
<span class="nt">&lt;/model&gt;</span>
</pre></div>
</div>
<p>And the model of the Flagellum:</p>
<p><strong>Flagellum.xml</strong></p>
<div class="highlight-xml notranslate"><div class="highlight"><pre><span></span> <span class="nt">&lt;model</span> <span class="na">inter_gene_max_space=</span><span class="s">&quot;20&quot;</span> <span class="na">min_mandatory_genes_required=</span><span class="s">&quot;9&quot;</span> <span class="na">min_genes_required=</span><span class="s">&quot;10&quot;</span> <span class="na">multi_loci=</span><span class="s">&quot;1&quot;</span> <span class="na">vers=</span><span class="s">&quot;2.0&quot;</span><span class="nt">&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctJ_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctN_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctQ_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctR_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctS_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctT_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctU_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_sctV_FLG&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_flgB&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_flgC&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;Flg_fliE&quot;</span> <span class="na">presence=</span><span class="s">&quot;mandatory&quot;</span><span class="nt">/&gt;</span>
   <span class="nt">&lt;gene</span> <span class="na">name=</span><span class="s">&quot;T3SS_sctC&quot;</span> <span class="na">presence=</span><span class="s">&quot;forbidden&quot;</span><span class="nt">/&gt;</span>
<span class="nt">&lt;/model&gt;</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="models_V1ToV2.html" class="btn btn-neutral float-right" title="Carrying models from v1 to v2" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="publish_package.html" class="btn btn-neutral" title="Publishing/sharing models" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2021, &#39;Institut Pasteur (Paris), CNRS&#39;

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    

  

  <script type="text/javascript" src="../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>