

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Frequently Asked Questions &mdash; MacSyFinder 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Modeller Guide" href="../modeler_guide/index.html" />
    <link rel="prev" title="MacSyFinder’s functioning" href="functioning.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MacSyFinder
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#running-macsyfinder">Running MacSyFinder</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsyfinder-functioning">MacSyFinder functioning</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Frequently Asked Questions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#how-to-report-an-issue">How to report an issue?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-cite-macsyfinder-and-published-macy-models">How to cite MacSyFinder and published macy-models?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-do-macsyfinder-command-lines-look-like">What do MacSyFinder command lines look like?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-search-mode-to-be-used">What search mode to be used?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-interpret-the-results-from-an-unordered-search">How to interpret the results from an <cite>unordered</cite> search?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-search-for-multiple-systems-at-once">How to search for multiple systems at once?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#when-can-the-option-previous-run-be-used">When can the option <cite>–previous-run</cite> be used?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#which-output-file-to-be-used-to-get-one-solution">Which output file to be used to get ONE solution?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#where-to-find-macsyfinder-models">Where to find MacSyFinder models?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-are-the-rules-for-options-precedence">What are the rules for options precedence?</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">User Guide</a> &raquo;</li>
        
      <li>Frequently Asked Questions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/user_guide/FAQ.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="frequently-asked-questions">
<span id="faq"></span><h1>Frequently Asked Questions<a class="headerlink" href="#frequently-asked-questions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="how-to-report-an-issue">
<h2>How to report an issue?<a class="headerlink" href="#how-to-report-an-issue" title="Permalink to this headline">¶</a></h2>
<p>If you encounter a problem while running MacSyFinder, please submit an issue on the dedicated page of the <a class="reference external" href="https://github.com/gem-pasteur/macsyfinder/issues">GitHub project</a></p>
<p>To ensure we have all elements to help, please provide:</p>
<ul class="simple">
<li><p>a concise description of the issue</p></li>
<li><p>the expected behavior VS observed one</p></li>
<li><p>the exact command-line used</p></li>
<li><p>the version of MacSyFinder used</p></li>
<li><p>the exact error message, and if applicable, the <cite>macsyfinder.log</cite> and <cite>macsyfinder.conf</cite> files</p></li>
<li><p>if applicable, an archive (or link to it) with the output files obtained</p></li>
<li><p>if possible, the smallest dataset there is to reproduce the issue</p></li>
<li><p>if applicable, this would also include the macsy-models (XML models plus HMM profiles) used (or precise version of the models if there are publicly available). Same as above, if possible, please provide the smallest set possible of models and HMM profiles.</p></li>
</ul>
<p>All these will definitely help us to help you! ;-)</p>
</div>
<div class="section" id="how-to-cite-macsyfinder-and-published-macy-models">
<span id="citations"></span><h2>How to cite MacSyFinder and published macy-models?<a class="headerlink" href="#how-to-cite-macsyfinder-and-published-macy-models" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pone.0110726">Abby et al. 2014</a>, <em>PLoS ONE</em> for the <strong>general principles of MacSyFinder</strong> (version 1), and the corresponding set of Cas systems (CasFinder, 1st version).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a>, <em>PLoS Genetics</em>, for the study of the evolutionary relationship between the T3SS and the bacterial flagellum, and how were designed the corresponding HMM protein profiles.</p></li>
<li><p><a class="reference external" href="https://www.nature.com/articles/srep23080">Abby et al. 2016</a>, <em>Scientific Reports</em>, for the description of bacterial protein secretion systems’ models (TXSScan: T1SS, T2SS, T5SS, T6SS, T9SS, Tad, T4P).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pbio.3000390">Denise et al. 2019</a>, <em>PLoS Biology</em>, for the description of type IV-filament super-family models (TFF-SF: T2SS, T4aP, T4bP, Com, Tad, archaeal T4P).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.ppat.1006525">Rendueles et al. 2017</a>, <em>PLoS Pathogens</em>, for the CapsuleFinder set of models.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1093/nar/gky425">Couvin, Bernheim et al. 2018</a>, <em>Nucleic Acids Research</em>, for the updated version of the set of Cas systems’ models, CasFinder.</p></li>
</ul>
</div>
<div class="section" id="what-do-macsyfinder-command-lines-look-like">
<span id="cmd-line-examples"></span><h2>What do MacSyFinder command lines look like?<a class="headerlink" href="#what-do-macsyfinder-command-lines-look-like" title="Permalink to this headline">¶</a></h2>
<p>Here are a few examples of command line formation:</p>
<p>To browse interactive help:</p>
<blockquote>
<div><p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">-h</span></code></p>
</div></blockquote>
<p>The minimal command line, to search all systems with models from the “TFF-SF” set of models (installed with <cite>macsydata</cite>):</p>
<blockquote>
<div><p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">genome.fasta</span> <span class="pre">--models</span> <span class="pre">TFF-SF</span> <span class="pre">all</span></code></p>
</div></blockquote>
<p>To search for several systems (ModelA and ModelB) from the “model_family” set of models that can be found in the “./my-models” folder:</p>
<blockquote>
<div><p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">genome.fasta</span> <span class="pre">--models</span> <span class="pre">model_family</span> <span class="pre">ModelA</span> <span class="pre">ModelB</span> <span class="pre">--models-dir</span> <span class="pre">./my-models</span></code></p>
</div></blockquote>
<p>To alter the search parameters and allow a maximal distance between components of 20 for the T2SS and 15 for the Tad pilus:</p>
<blockquote>
<div><p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">genome.fasta</span> <span class="pre">--models</span> <span class="pre">TFF-SF</span> <span class="pre">all</span> <span class="pre">--inter-gene-max-space</span> <span class="pre">T2SS</span> <span class="pre">20</span> <span class="pre">--inter-gene-max-space</span> <span class="pre">Tad</span> <span class="pre">15</span></code></p>
</div></blockquote>
<p>To alter the search parameters and allow the Tad pilus to be made of multiple loci:</p>
<blockquote>
<div><p><code class="code docutils literal notranslate"><span class="pre">macsyfinder</span> <span class="pre">--db-type</span> <span class="pre">ordered_replicon</span> <span class="pre">--sequence-db</span> <span class="pre">genome.fasta</span> <span class="pre">--models</span> <span class="pre">TFF-SF</span> <span class="pre">all</span> <span class="pre">--multi-loci</span> <span class="pre">Tad</span></code></p>
</div></blockquote>
<p>See also the <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">MacSyFinder Quick Start</span></a> section for more examples.</p>
</div>
<div class="section" id="what-search-mode-to-be-used">
<h2>What search mode to be used?<a class="headerlink" href="#what-search-mode-to-be-used" title="Permalink to this headline">¶</a></h2>
<p>Depending on the type of dataset you have, you will have to adapt MacSyFinder’s search mode.</p>
<ul class="simple">
<li><p>If you have a fasta file from a complete genome where <strong>proteins are ordered</strong> according to the corresponding genes’ order along the replicon, your dataset is entitled to the most powerful search mode (see below): <cite>ordered_replicon</cite> and use the following option <cite>–db-type ordered_replicon</cite>.</p></li>
<li><p>If you have a fasta file of proteins with <strong>no sense of the order</strong> of the corresponding genes along the chromosome(s) or replicon(s), you will have to use the <cite>unordered</cite> search mode with the following option: <cite>–db-type unordered</cite></p></li>
<li><p>If you have <strong>multiple ordered replicons</strong> to analyse at once, you can follow the <cite>Gembase</cite> convention to name the proteins in the fasta file, so that the original replicons can be assessed from their name: <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">see here for a description</span></a>.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>When the <strong>gene order is known</strong> (<cite>ordered_replicon</cite> search mode) the power of the analysis is <strong>maximal</strong>, since both the genomic content and context are taken into account for the search.</p></li>
<li><p>When the <strong>gene order is unknown</strong> (<cite>unordered</cite> search mode) the power of the analysis is more <strong>limited</strong> since the presence of systems can only be suggested on the basis of the quorum of components - and not based on genomic context information.</p></li>
</ul>
</div>
<p>More on command-line options <a class="reference internal" href="input.html#command-line-label"><span class="std std-ref">here</span></a> and on MacSyFinder’s functioning <a class="reference internal" href="functioning.html#functioning"><span class="std std-ref">here</span></a>.</p>
</div>
<div class="section" id="how-to-interpret-the-results-from-an-unordered-search">
<h2>How to interpret the results from an <cite>unordered</cite> search?<a class="headerlink" href="#how-to-interpret-the-results-from-an-unordered-search" title="Permalink to this headline">¶</a></h2>
<p>As mentioned above, in the <cite>unordered</cite> search mode, the inference of a system’s presence is only based on the list of components found in the protein dataset.
Thus, the kind of search specificity provided when using the genomic context (components next to each other are more likely to be part of a same system’s occurrence) is not within reach.</p>
<p>In the <cite>unordered</cite> search mode, the number of proteins selected as system’s components (based on the filtering of HMM profiles’ similarity search) is reported.
We decided to report all kinds of system’s components, including the <cite>forbidden</cite> ones in order to raise awareness of the user -&gt; even if all constraints are met for the system’s inference (here, the quorum: minimal number of components), it cannot be excluded that a <cite>forbidden</cite> component would lie next to the <em>bona fide</em> components (<cite>mandatory</cite> and <cite>accessory</cite> ones) in the genome…</p>
<p>In the end, the <cite>unordered</cite> search mode provides an idea as to whether the <strong>genetic potential</strong> for a given system is found in the set of proteins analysed, with no attempt to assign proteins to particular systems’ occurrences, nor guarantee as to whether <cite>forbidden</cite> components should be considered for the potential occurrences.</p>
</div>
<div class="section" id="how-to-search-for-multiple-systems-at-once">
<h2>How to search for multiple systems at once?<a class="headerlink" href="#how-to-search-for-multiple-systems-at-once" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>It is possible to search for only some systems from a macsy-model package. In this case, the command-line should be formed as follows:</p></li>
</ul>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>macsyfinder --models TXSS Flagellum T2SS --sequence-db mygenomes.fasta --db-type gembase
</pre></div>
</div>
<p>This would run the search of the systems “Flagellum” and “T2SS” in the dataset “mygenomes.fasta”.</p>
<ul class="simple">
<li><p>To run the search of all the models contained in a macsy-model package, use the following:</p></li>
</ul>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>macsyfinder --models TXSS all --sequence-db mygenomes.fasta --db-type gembase
macsyfinder --models CRISPRCas all --sequence-db mygenomes.fasta --db-type gembase
macsyfinder --models CRISPRCas/typing all --sequence-db mygenomes.fasta --db-type gembase
</pre></div>
</div>
<p>You can see that the <cite>all</cite> keyword can not only be applied to an entire macsy-model package and its entire hierarchy, but can also be ran on all the systems from a macsy-model sub-directory.</p>
</div>
<div class="section" id="when-can-the-option-previous-run-be-used">
<h2>When can the option <cite>–previous-run</cite> be used?<a class="headerlink" href="#when-can-the-option-previous-run-be-used" title="Permalink to this headline">¶</a></h2>
<p>The option <cite>–previous-run</cite> enables to avoid running the HMM profile search and the hits extraction when the set of systems to search and the replicons to analyse are exactly the same between runs.
This enables to alter the features of the systems to be searched for, i.e. basically any feature found in the XML file of the corresponding models:</p>
<ul class="simple">
<li><p>the maximal distance allowed between components to be considered as part of a same locus <cite>–inter-gene-max-space</cite></p></li>
<li><p>the minimal number of components to be found to infer a full system <cite>–min-mandatory-genes-required</cite> and <cite>–min-genes-required</cite></p></li>
<li><p>the general genomic architecture of the system <cite>–multi-loci</cite></p></li>
</ul>
<p>This also means that there are a number of options that are incompatible with  <cite>–previous-run</cite>, including:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>--config, --sequence-db, --profile-suffix, --res-extract-suffix, --e-value-res, --db-type, --hmmer
</pre></div>
</div>
</div>
<div class="section" id="which-output-file-to-be-used-to-get-one-solution">
<h2>Which output file to be used to get ONE solution?<a class="headerlink" href="#which-output-file-to-be-used-to-get-one-solution" title="Permalink to this headline">¶</a></h2>
<p>Since version 2 of MacSyFinder, a combinatorial exploration of the possible sets of systems is performed.
A scoring scheme has been set up to differentiate between solutions,
in order to provide the user with the most complete set of systems as possible given the searched models.
This score is maximal for the “best solution”. This also means that some solutions might get the same maximal score.
In this case, one can wonder how to find all the equivalent solutions, and an other,
how to simply pick one solution among the best, whichever it is.
We thus propose several kind of <a class="reference internal" href="outputs.html#ordered-outputs"><span class="std std-ref">output files</span></a>.</p>
<ul class="simple">
<li><p>All equivalent best solutions are found in the <cite>all_best_solutions.tsv</cite> file.</p></li>
<li><p>One best solution is given in the <cite>best_solution.tsv</cite> file.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For those more familiar with the output files from MacSyFinder v1, the file <cite>best_solution.tsv</cite> is the closest from
the previous output file <cite>macsyfinder.report</cite>.</p>
</div>
</div>
<div class="section" id="where-to-find-macsyfinder-models">
<h2>Where to find MacSyFinder models?<a class="headerlink" href="#where-to-find-macsyfinder-models" title="Permalink to this headline">¶</a></h2>
<p>Since version 2, there is a tool to enable the download and installation of published models from a repository: the <cite>macsydata</cite> tool.</p>
<p>See <a class="reference internal" href="installation.html#macsydata"><span class="std std-ref">here for details</span></a> on how to use it.</p>
</div>
<div class="section" id="what-are-the-rules-for-options-precedence">
<h2>What are the rules for options precedence?<a class="headerlink" href="#what-are-the-rules-for-options-precedence" title="Permalink to this headline">¶</a></h2>
<p>MacSyFinder offers many ways to parametrize the systems’ search: through the command-line, through various configuration files (for the models, for the run, etc…). It offers a large control over the search engine. But it also means you can get lost in configuration. ;-)</p>
<p>Here is a recap of the rules for options precedence. In a general manner, the command line always wins.</p>
<p>The precedence rules between the different levels of configuration are:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>system &lt; home &lt; model &lt; project &lt; --cfg-file | --previous-run &lt; command line options
</pre></div>
</div>
<ul class="simple">
<li><p><strong>system</strong>: the <cite>macsyfinder.conf</cite> file either in /etc/macsyfinder/ or in virtalenv/etc/macsyfinder/
in case of a <em>virtualenv</em> this configuration affects only the MacSyFinder version installed in this virtualenv</p></li>
<li><p><strong>home</strong>:  the <cite>~/.macsyfinder/macsyfinder.conf</cite> file</p></li>
<li><p><strong>model</strong>: the <cite>model_conf.xml</cite> file at the root of the model package</p></li>
<li><p><strong>project</strong>: the <cite>macsyfinder.conf</cite> file found in the directory where the <cite>macsyfinder</cite> command was run</p></li>
<li><p><strong>cfgfile</strong>: any configuration file specified by the user on the command line (conflicts with the <cite>–previous-run</cite> option)</p></li>
<li><p><strong>previous-run</strong>: the <cite>macsyfinder.conf</cite> file found in the results directory of the previous run (conflicts with the <cite>–cfg-file</cite> option)</p></li>
<li><p><strong>command line</strong>: any option specified directly in the command line</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../modeler_guide/index.html" class="btn btn-neutral float-right" title="Modeller Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="functioning.html" class="btn btn-neutral" title="MacSyFinder’s functioning" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2021, &#39;Institut Pasteur (Paris), CNRS&#39;

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    

  

  <script type="text/javascript" src="../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>