

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Input and Options of MacSyFinder &mdash; MacSyFinder 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Gembase format" href="gembase_convention.html" />
    <link rel="prev" title="MacSyFinder Quick Start" href="quickstart.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MacSyFinder
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#running-macsyfinder">Running MacSyFinder</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="new_v2.html">What’s new in MacSyFinder v2?</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html">MacSyFinder Quick Start</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Input and Options of MacSyFinder</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input-dataset">Input dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#command-line-options">Command-line options</a></li>
<li class="toctree-l4"><a class="reference internal" href="#configuration-file">Configuration file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#in-house-input-files">In-house input files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="outputs.html">Output format</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsyfinder-functioning">MacSyFinder functioning</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">User Guide</a> &raquo;</li>
        
      <li>Input and Options of MacSyFinder</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/user_guide/input.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="input-and-options-of-macsyfinder">
<span id="input"></span><h1>Input and Options of MacSyFinder<a class="headerlink" href="#input-and-options-of-macsyfinder" title="Permalink to this headline">¶</a></h1>
<div class="section" id="input-dataset">
<span id="input-dataset-label"></span><h2>Input dataset<a class="headerlink" href="#input-dataset" title="Permalink to this headline">¶</a></h2>
<p>The input dataset must be a set of protein sequences in <strong>Fasta format</strong> (see <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">http://en.wikipedia.org/wiki/FASTA_format</a>).</p>
<p>The <a class="reference internal" href="#config-base-label"><span class="std std-ref">base section</span></a> in the configuration file (see <a class="reference internal" href="#config-definition-label"><span class="std std-ref">Configuration file</span></a>)
can be used to specify <strong>the path</strong> and the <strong>type of dataset</strong> to deal with,
as well as the <cite>–sequence_db</cite> and <cite>–db_type</cite> parameters respectively,
described in the <a class="reference internal" href="#command-line-label"><span class="std std-ref">Command-line options</span></a> (see <a class="reference internal" href="#cmd-input-label"><span class="std std-ref">Input options</span></a>).</p>
<blockquote>
<div><p>Four types of protein datasets are supported:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>unordered</em> : a set of sequences corresponding to a complete genome
(<em>e.g.</em> an unassembled complete genome)</p></li>
<li><p><em>ordered_replicon</em> : a set of sequences corresponding to an ordered complete replicon
(<em>e.g.</em> an assembled complete genome)</p></li>
<li><p><em>gembase</em> : a set of multiple ordered replicons, which format follows the convention described
in <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">Gembase format</span></a>.</p></li>
</ul>
</div></blockquote>
</div></blockquote>
<p>For “ordered” (“ordered_replicon” or “gembase”) datasets only,
MacSyFinder can take into account the <strong>shape of the genome</strong>: “linear”,
or “circular” for detection. The default is set to “circular”.</p>
<blockquote>
<div><p>This can be set with the <cite>–replicon_topology</cite> parameter from <a class="reference internal" href="#command-line-label"><span class="std std-ref">Command-line options</span></a>
(see <a class="reference internal" href="#cmd-input-label"><span class="std std-ref">Input options</span></a>),
or in the configuration in the <a class="reference internal" href="#config-base-label"><span class="std std-ref">base section</span></a>.</p>
<p>With the “gembase” format, it is possible to specify a topology per replicon with a topology file
(see <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">Gembase format</span></a> and <a class="reference internal" href="gembase_convention.html#topology-files"><span class="std std-ref">Topology files</span></a>).</p>
</div></blockquote>
</div>
<div class="section" id="command-line-options">
<span id="command-line-label"></span><h2>Command-line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<p>Optional arguments:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>-h, --help            Show the help message and exit

-m [MODELS [MODELS ...]], --models [MODELS [MODELS ...]]
                      The models to search. The --models option can be set several times.&#39;
                      For each --models options the first element must be the name of family models,
                      followed by the name of the models.
                      If the name &#39;all&#39; is in the list all models from the family will be searched.&#39;
                      &#39;--models TXSS Flagellum T2SS&#39;
                                means MSF will search for models TXSS/Flagellum and TXSS/T2SS
                      &#39;--models TXSS all&#39;
                                means for all models found in model package TXSS
                      &#39;--models CRIPRcas/subtyping all&#39;
                               means MSF will search for all models described in the CRISPRCas/subtyping subfamily.
                      (required unless --previous-run is set)
</pre></div>
</div>
<p id="cmd-input-label">Input dataset options:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>--sequence-db SEQUENCE_DB
                      Path to the sequence dataset in fasta format.
                      (required unless --previous-run is set)
--db-type {ordered_replicon,gembase,unordered}
                      The type of dataset to deal with. &quot;unordered&quot; corresponds
                      to a non-assembled genome,
                      &quot;ordered_replicon&quot; to an assembled genome,
                      and &quot;gembase&quot; to a set of replicons where sequence identifiers
                      follow this convention: &quot;&gt;RepliconName SequenceID&quot;.
                      (required unless --previous-run is set)
--replicon-topology {linear,circular}
                      The topology of the replicons
                      (this option is meaningful only if the db_type is
                      &#39;ordered_replicon&#39; or &#39;gembase&#39;.
--topology-file TOPOLOGY_FILE
                      Topology file path. The topology file allows to specify a topology
                      (linear or circular) for each replicon (this option is meaningful only if
                      the db_type is &#39;ordered_replicon&#39; or &#39;gembase&#39;.
                      A topology file is a tabular file with two columns:
                      the 1st is the replicon name, and the 2nd the corresponding topology:
                      &quot;RepliconA      linear&quot;
--idx                 Forces to build the indexes for the sequence dataset even
                      if they were previously computed and present at the dataset location.
                      (default: False)
</pre></div>
</div>
<p id="system-detect-options">Systems detection options:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>--inter-gene-max-space INTER_GENE_MAX_SPACE INTER_GENE_MAX_SPACE
                      Co-localization criterion: maximum number of components non-matched by a
                      profile allowed between two matched components
                      for them to be considered contiguous.
                      Option only meaningful for &#39;ordered&#39; datasets.
                      The first value must match to a model, the second to a number of components.
                      This option can be repeated several times:
                          &quot;--inter-gene-max-space TXSS/T2SS 12 --inter-gene-max-space TXSS/Flagellum 20
--min-mandatory-genes-required MIN_MANDATORY_GENES_REQUIRED MIN_MANDATORY_GENES_REQUIRED
                      The minimal number of mandatory genes required for model assessment.
                      The first value must correspond to a model fully qualified name, the second value to an integer.
                      This option can be repeated several times:
                          &quot;--min-mandatory-genes-required TXSS/T2SS 15 --min-mandatory-genes-required TXSS/Flagellum 10&quot;
--min-genes-required MIN_GENES_REQUIRED MIN_GENES_REQUIRED
                      The minimal number of genes required for model assessment &quot;
                      (includes both &#39;mandatory&#39; and &#39;accessory&#39; components).
                      The first value must correspond to a model fully qualified name, the second value to an integer.
                      This option can be repeated several times:
                          &quot;--min-genes-required TXSS/T2SS 15 --min-genes-required TXSS/Flagellum 10
--max-nb-genes MAX_NB_GENES MAX_NB_GENES
                      The maximal number of genes to consider a system as full.
                      The first value must correspond to a model name, the second value to an integer.
                      This option can be repeated several times:
                          &quot;--max-nb-genes TXSS/T2SS 5 --max-nb-genes TXSS/Flagellum 10&quot;
--multi-loci MULTI_LOCI
                      Specifies if the system can be detected as a &#39;scattered&#39; system.
                      The models are specified as a comma separated list of fully qualified name
                          &quot;--multi-loci model_familyA/model_1,model_familyB/model_2&quot;
</pre></div>
</div>
<p id="hmmer-options">Options for Hmmer execution and hits filtering:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span> --hmmer HMMER         Path to the hmmsearch program.
                       If it is not specify rely on the PATH
                       (default: hmmsearch)
--e-value-search E_VALUE_SEARCH
                       Maximal e-value for hits to be reported during hmmsearch search.
                       By default MF set per profile threshold for hmmsearch run (--cut_ga option)
                       for profiles containing the GA bit score threshold.
                       If a profile does not contains the GA bit score the --e-value-search (-E in hmmsearch)
                       is applied to this profile.
                       To applied the --e-value-search to all profiles use the --no-cut-ga option.
                       (default: 0.1)
 --no-cut-ga           By default the Mf try to applied a threshold per profile by using the
                       hmmer -cut-ga option. This is possible only if the Ga bit score is present in the profile otherwise MF switch to use the
                       the --e-value-search (-E in hmmsearch).
                       If this option is set the --e-value-search option is used for all profiles regardless
                       the presence of the a GA bit score in the profiles.
                       (default: False)
 --i-evalue-sel I_EVALUE_SEL
                       Maximal independent e-value for Hmmer hits to be selected for system detection.
                       (default:0.001)
 --coverage-profile COVERAGE_PROFILE
                       Minimal profile coverage required in the hit alignment to allow
                       the hit selection for system detection.
                       (default: 0.5)
</pre></div>
</div>
<p id="score-options">Options for clusters and systems’ scoring:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>--mandatory-weight MANDATORY_WEIGHT
                      the weight (score) of a mandatory component when scoring clusters
                      (default:1.0)
--accessory-weight ACCESSORY_WEIGHT
                      the weight (score) of an accessory component when scoring clusters
                      (default:0.5)
--exchangeable-weight EXCHANGEABLE_WEIGHT
                      the weight modifier for the score of a component that is exchangeable
                      (default:0.8)
--redundancy-penalty REDUNDANCY_PENALTY
                      the weight modifier for the score of a component that is already present in another cluster
                      (default:1.5)

--loner-multi-system-weight LONER_MULTI_SYSTEM_WEIGHT
                      the weight modifier for the score of a component that is `loner` and `multi-system` at the same time
                      (default:0.7)
</pre></div>
</div>
<p id="path-options">Path options:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>--models-dir MODELS_DIR
                      specify the path to the models if the models are not installed in the canonical place.
                      It gather definitions (xml files) and hmm profiles in a specific
                      structure. A directory with the name of the model with at least two directories
                      profiles&quot; which contains all hmm profile for gene describe in definitions and
                      models&quot; which contains either xml file of definitions or subdirectories
                      to organize the model in subsystems.
-o OUT_DIR, --out-dir OUT_DIR
                      Path to the directory where to store results.
                      if out-dir is specified res-search-dir will be ignored.
--res-search-suffix RES_SEARCH_SUFFIX
                      The suffix to give to Hmmer raw output files. (default: .search_hmm.out)
--res-extract-suffix RES_EXTRACT_SUFFIX
                      The suffix to give to filtered hits output files. (default: .res_hmm_extract)
--profile-suffix PROFILE_SUFFIX
                      The suffix of profile files. For each &#39;Gene&#39; element, the corresponding profile is
                      searched in the &#39;profile_dir&#39;, in a file which name is based on the
                      Gene name + the profile suffix.
                      For instance, if the Gene is named &#39;gspG&#39; and the suffix is &#39;.hmm3&#39;,
                      then the profile should be placed at the specified location
                      and be named &#39;gspG.hmm3&#39;
                      (default: .hmm)
</pre></div>
</div>
<p>General options:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>-w WORKER, --worker WORKER
                      Number of workers to be used by MacSyFinder.
                      In the case the user wants to run MacSyFinder in a multi-thread mode.
                      (0 mean all cores will be used).
                      (default: 1)
-v, --verbosity       Increases the verbosity level. There are 4 levels:
                      Error messages (default), Warning (-v), Info (-vv) and Debug.(-vvv)
--mute                mute the log on stdout.
                      (continue to log on macsyfinder.log)
                      (default: False)
--version             show program&#39;s version number and exit
-l, --list-models     display the all models installed in generic location and quit.
--cfg-file CFG_FILE   Path to a MacSyFinder configuration file to be used.
--previous-run PREVIOUS_RUN
                      Path to a previous MacSyFinder run directory.
                      It allows to skip the Hmmer search step on same dataset,
                      as it uses previous run results and thus parameters regarding Hmmer detection.
                      The configuration file from this previous run will be used.
                      Conflict with options
                          --config, --sequence-db, --profile-suffix, --res-extract-suffix, --e-value-res, --db-type, --hmmer
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For some command line examples, have a look <a class="reference internal" href="FAQ.html#cmd-line-examples"><span class="std std-ref">here</span></a>, or at the <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">MacSyFinder Quick Start</span></a> section.</p>
</div>
</div>
<div class="section" id="configuration-file">
<span id="config-definition-label"></span><h2>Configuration file<a class="headerlink" href="#configuration-file" title="Permalink to this headline">¶</a></h2>
<p>Options to run MacSyFinder can be specified in a configuration file.
The <a class="reference internal" href="../developer_guide/api/config.html#config"><span class="std std-ref">Config object</span></a> handles all configuration options for MacSyFinder.
Three locations are parsed to find configuration files:</p>
<blockquote>
<div><ul class="simple">
<li><p>$PREFIX/etc/macsyfinder/macsyfinder.conf</p></li>
<li><p>$(HOME)/.macsyfinder/macsyfinder.conf</p></li>
<li><p>./macsyfinder.conf</p></li>
</ul>
</div></blockquote>
<p>Moreover these three locations options can be passed on the command-line.</p>
<p>Each file can define options, and in the end all options are integrated. If an option is specified several times:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The precedence rules from the least to the most important priority are:</p>
<p>$PREFIX/etc/macsyfinder/macsyfinder.conf &lt; $(HOME)/.macsyfinder/macsyfinder.conf &lt; ./macsyfinder.conf &lt; “command-line” options</p>
</div>
<p>This means that command-line options will always bypass those from the configuration files. In the same flavor,
options altering the definition of systems found in the command-line or the configuration file will always
overwhelm values from systems’ <a class="reference internal" href="../modeler_guide/modeling.html#model-definition-grammar-label"><span class="std std-ref">XML definition files</span></a>.</p>
<p>The configuration files must follow the Python “ini” file syntax.
The <a class="reference internal" href="../developer_guide/api/config.html#config"><span class="std std-ref">Config object</span></a> provides some default values and performs some validations of the values.</p>
<p>In MacSyFinder, six sections are defined and stored by default in the configuration file:</p>
<blockquote>
<div><blockquote id="config-base-label">
<div><ul>
<li><p><strong>base</strong> : all information related to the protein dataset under study</p>
<ul>
<li><p><em>sequence_db</em> : the path to the dataset in Fasta format (<em>no default value</em>)</p></li>
<li><p><em>db_type</em> : the type of dataset to handle, four types are supported:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>unordered</em> : a set of sequences corresponding to a complete replicon
(<em>e.g.</em> an unassembled complete genome)</p></li>
<li><p><em>ordered_replicon</em> : a set of sequences corresponding to a complete replicon ordered
(<em>e.g.</em> an assembled complete genome)</p></li>
<li><p><em>gembase</em> : a set of multiple ordered replicons.</p></li>
</ul>
</div></blockquote>
<p>(<em>no default value</em>)</p>
</li>
<li><p><em>replicon_topology</em> : the topology of the replicon under study.
Two topologies are supported: ‘linear’ and ‘circular’ (<em>default</em> = ‘circular’).
This option will be ignored if the dataset type is not ordered (<em>i.e.</em> “unordered_replicon” or “unordered”).</p></li>
</ul>
</li>
<li><p><strong>models</strong>
* list of models to search in replicon</p></li>
<li><p><strong>models_opt</strong></p>
<ul class="simple">
<li><p><em>inter_gene_max_space</em> = list of models’ fully qualified names and integer separated by spaces (see example below).
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>min_mandatory_genes_required</em> = list of models’ fully qualified name and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>min_genes_required</em> = list of models’ fully qualified name and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
<li><p><em>max_nb_genes</em> = list of models’ fully qualified names and integer separated by spaces.
These values will supersede the values found in the model definition file.</p></li>
</ul>
</li>
<li><p><strong>hmmer</strong></p>
<ul class="simple">
<li><p><em>hmmer_exe</em> (default= <em>hmmsearch</em> )</p></li>
<li><p><em>e_value_res</em> = (default= <em>1</em> )</p></li>
<li><p><em>i_evalue_sel</em> = (default= <em>0.5</em> )</p></li>
<li><p><em>coverage_profile</em> = (default= <em>0.5</em> )</p></li>
</ul>
</li>
<li><p><strong>score_opt</strong></p>
<ul class="simple">
<li><p><em>mandatory_weight</em> (default= <em>1.0</em>)</p></li>
<li><p><em>accessory_weight</em> (default= <em>0.5</em>)</p></li>
<li><p><em>exchangeable_weight</em> (default= <em>0.8</em>)</p></li>
<li><p><em>redundancy_penalty</em> (default= <em>1.5</em>)</p></li>
<li><p><em>loner_multi_system_weight</em> (default= <em>0.7</em>)</p></li>
</ul>
</li>
<li><p><strong>directories</strong></p>
<ul class="simple">
<li><p><em>res_search_dir</em> = (default= <em>./datatest/res_search</em> )</p></li>
<li><p><em>res_search_suffix</em> = (default= <em>.search_hmm.out</em> )</p></li>
<li><p><em>models_dir</em> = (default= <em>./models</em> )</p></li>
<li><p><em>res_extract_suffix</em> = (default= <em>.res_hmm_extract</em> )</p></li>
</ul>
</li>
<li><p><strong>general</strong></p>
<ul>
<li><dl>
<dt><em>log_level</em>: (default= <em>debug</em> ) This corresponds to an integer code:</dt><dd><table class="docutils align-default">
<colgroup>
<col style="width: 38%" />
<col style="width: 62%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>Numeric value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CRITICAL</p></td>
<td><p>50</p></td>
</tr>
<tr class="row-odd"><td><p>ERROR</p></td>
<td><p>40</p></td>
</tr>
<tr class="row-even"><td><p>WARNING</p></td>
<td><p>30</p></td>
</tr>
<tr class="row-odd"><td><p>INFO</p></td>
<td><p>20</p></td>
</tr>
<tr class="row-even"><td><p>DEBUG</p></td>
<td><p>10</p></td>
</tr>
<tr class="row-odd"><td><p>NOTSET</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</dd>
</dl>
</li>
<li><p><em>log_file</em> = (default = macsyfinder.log in directory of the run)</p></li>
</ul>
</li>
</ul>
</div></blockquote>
</div></blockquote>
<p>Example of a configuration file</p>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span> <span class="k">[base]</span>
 <span class="na">prefix</span> <span class="o">=</span> <span class="s">/path/to/macsyfinder/home/</span>
 <span class="na">file</span> <span class="o">=</span> <span class="s">%(prefix)s/data/base/prru_psae.001.c01.fasta</span>
 <span class="na">db_type</span> <span class="o">=</span> <span class="s">gembase</span>
 <span class="na">replicon_topology</span> <span class="o">=</span> <span class="s">circular</span>

 <span class="k">[models]</span>
 <span class="na">models_1</span> <span class="o">=</span> <span class="s">TFF-SF_final all</span>

 <span class="k">[models_opt]</span>
 <span class="na">inter_gene_max_space</span> <span class="o">=</span> <span class="s">TXSS/T2SS 22 TXSS/Flagellum 44</span>
 <span class="na">min_mandatory_genes_required</span> <span class="o">=</span> <span class="s">TXSS/T2SS 6 TXSS/Flagellum 4</span>
 <span class="na">min_genes_required</span> <span class="o">=</span> <span class="s">TXSS/T2SS 8 TXSS/Flagellum 4</span>
 <span class="na">max_nb_genes</span> <span class="o">=</span> <span class="s">TXSS/T2SS 12 TXSS/Flagellum 8</span>

 <span class="k">[hmmer]</span>
 <span class="na">hmmer</span> <span class="o">=</span> <span class="s">hmmsearch</span>
 <span class="na">e_value_res</span> <span class="o">=</span> <span class="s">1</span>
 <span class="na">i_evalue_sel</span> <span class="o">=</span> <span class="s">0.5</span>
 <span class="na">coverage_profile</span> <span class="o">=</span> <span class="s">0.5</span>

 <span class="k">[score_opt]</span>
 <span class="na">mandatory_weight</span> <span class="o">=</span> <span class="s">1.0</span>
 <span class="na">accessory_weight</span> <span class="o">=</span> <span class="s">0.5</span>
 <span class="na">exchangeable_weight</span> <span class="o">=</span> <span class="s">0.8</span>
 <span class="na">redundancy_penalty</span> <span class="o">=</span> <span class="s">1.5</span>
 <span class="na">loner_multi_system_weight</span> <span class="o">=</span> <span class="s">0.7</span>

 <span class="k">[directories]</span>
 <span class="na">prefix</span> <span class="o">=</span> <span class="s">/path/to/macsyfinder/home/</span>
 <span class="na">data_dir</span> <span class="o">=</span> <span class="s">%(prefix)s/data/</span>
 <span class="na">res_search_dir</span> <span class="o">=</span> <span class="s">%(prefix)s/dataset/res_search/</span>
 <span class="na">res_search_suffix</span> <span class="o">=</span> <span class="s">.raw_hmm</span>
 <span class="na">models_dir</span> <span class="o">=</span> <span class="s">%(data_dir)/data/models</span>
 <span class="na">profile_suffix</span> <span class="o">=</span> <span class="s">.fasta-aln.hmm</span>
 <span class="na">res_extract_suffix</span> <span class="o">=</span> <span class="s">.res_hmm</span>

<span class="k">[general]</span>
<span class="na">log_level</span> <span class="o">=</span> <span class="s">debug</span>
<span class="na">worker</span> <span class="o">=</span> <span class="s">4</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>After a run, the corresponding configuration file (“macsyfinder.conf”) is generated as a (re-usable)
output file that stores every options used in the run.
It is stored in the results’ directory (see <a class="reference internal" href="outputs.html#outputs"><span class="std std-ref">the output section</span></a>).</p>
</div>
</div>
<div class="section" id="in-house-input-files">
<h2>In-house input files<a class="headerlink" href="#in-house-input-files" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="gembase_convention.html">Gembase format</a></li>
<li class="toctree-l1"><a class="reference internal" href="gembase_convention.html#topology-files">Topology files</a></li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="gembase_convention.html" class="btn btn-neutral float-right" title="Gembase format" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="quickstart.html" class="btn btn-neutral" title="MacSyFinder Quick Start" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2021, &#39;Institut Pasteur (Paris), CNRS&#39;

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    

  

  <script type="text/javascript" src="../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>