

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Output format &mdash; MacSyFinder 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Macromolecular models" href="model_definition.html" />
    <link rel="prev" title="Gembase format" href="gembase_convention.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MacSyFinder
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#running-macsyfinder">Running MacSyFinder</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="new_v2.html">What’s new in MacSyFinder v2?</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html">MacSyFinder Quick Start</a></li>
<li class="toctree-l3"><a class="reference internal" href="input.html">Input and Options of MacSyFinder</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Output format</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hmmer-results-output-files">Hmmer results’ output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#output-files-for-the-ordered-replicon-s-search-modes">Output files for the “ordered replicon(s)” search modes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#output-files-for-the-unordered-replicon-search-mode">Output files for the “unordered replicon” search mode</a></li>
<li class="toctree-l4"><a class="reference internal" href="#logs-and-configuration-files">Logs and configuration files</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsyfinder-functioning">MacSyFinder functioning</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">User Guide</a> &raquo;</li>
        
      <li>Output format</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/user_guide/outputs.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="output-format">
<span id="outputs"></span><h1>Output format<a class="headerlink" href="#output-format" title="Permalink to this headline">¶</a></h1>
<p>MacSyFinder provides different types of output files. At each run, MacSyFinder creates a new folder,
whose name is based on a fixed prefix and a random suffix, for instance “macsyfinder-20130128_08-57-46”.
MacSyFinder output files are stored in this run-specific folder.</p>
<div class="section" id="hmmer-results-output-files">
<span id="hmmer-outputs-label"></span><h2>Hmmer results’ output files<a class="headerlink" href="#hmmer-results-output-files" title="Permalink to this headline">¶</a></h2>
<p>Raw Hmmer outputs are provided, as long with processed tabular outputs that include hits filtered as
specified by the user. For instance, the Hmmer search for SctC homologs with the corresponding profile
will result in the creation of two output files: “sctC.search_hmm.out” for the raw HMMER output file and
“sctC.res_hmm_extract” for the output file after processing/filtering of the HMMER results by MacSyFinder.</p>
<p>The processed output file “sctC.res_hmm_extract” recalls on the first lines the parameters used for
hits filtering and relevant information on the matches, as in this example:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># gene: sctC extract from /Users/bob/macsyfinder_results/
      macsyfinder-20130128_08-57-46/sctC.search_hmm.out hmm output
# profile length= 544
# i_evalue threshold= 0.001000
# coverage threshold= 0.500000
# hit_id replicon_name position_hit hit_sequence_length gene_name gene_system i_eval score
      profile_coverage sequence_coverage begin end
PSAE001c01_006940       PSAE001c01      3450    803     sctC    T3SS    1.1e-41 141.6
      0.588235  0.419676        395     731
PSAE001c01_018920       PSAE001c01      4634    776     sctC    T3SS    9.2e-48 161.7
      0.976103  0.724227        35      596
PSAE001c01_031420       PSAE001c01      5870    658     sctC    T3SS    2.7e-52 176.7
      0.963235  0.844985        49      604
PSAE001c01_051090       PSAE001c01      7801    714     sctC    T3SS    1.9e-46 157.4
      0.571691  0.463585        374     704
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Each tabular output file contains a header line describing each column in the output.</p>
</div>
</div>
<div class="section" id="output-files-for-the-ordered-replicon-s-search-modes">
<span id="ordered-outputs"></span><h2>Output files for the “ordered replicon(s)” search modes<a class="headerlink" href="#output-files-for-the-ordered-replicon-s-search-modes" title="Permalink to this headline">¶</a></h2>
<p>These output files are provided when MacSyFinder search proceeds on a set of proteins that are deemed to follow the order of their genes on replicons.
This corresponds to the two search modes <em>gembase</em> and <em>ordered_replicon</em>.</p>
<div class="section" id="systems-detection-results">
<h3>Systems detection results<a class="headerlink" href="#systems-detection-results" title="Permalink to this headline">¶</a></h3>
<p>Different types of output files are provided, human-readable files “.txt”, and tabulated files “.tsv”. For the latter,
headers are provided with the content of the lines in the file.</p>
<blockquote>
<div><ul>
<li><p><strong>best_solution.tsv</strong> - This file contains the <strong>best solution found by MacSyFinder</strong> in terms of systems detected,
under the form of a per-component, tabulated report file. A <strong>solution</strong> consists in a set of compatible systems (no components’ overlap allowed).
If multiple solutions showed a maximal score, the solution maximizing</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>the number of components in systems and</p></li>
<li><p>the number of systems detected is proposed.</p></li>
</ol>
</div></blockquote>
<p>To see potential other best solutions (in case several obtained the same highest score), see file <cite>all_best_solutions.tsv</cite>.</p>
<p>To see all possible, candidate systems without further processing, see files <cite>all_systems.txt</cite> and <cite>all_systems.tsv</cite>.</p>
<p>The <cite>best_solution.tsv</cite> file is the most similar to former V1 file <cite>macsyfinder.report</cite>.</p>
</li>
<li><p><strong>all_systems.txt</strong> - This file describes the search process of all possible candidate systems given the definitions in systems’ models -
without processing of the potential overlaps between candidate systems. This set of possible candidate systems are also given
under the form of a tabulated file in <cite>all_systems.tsv</cite>.</p></li>
<li><p><strong>rejected_clusters.txt</strong> - This file lists candidate clusters of systems’ components that were rejected by
MacSyFinder during the search process, and were thus not assigned to a candidate system.</p></li>
<li><p><strong>all_best_solutions.tsv</strong> - This file contains all possible best solutions under the form of a per-component, tabulated report file.
To retrieve a single best solution as proposed by MacSyFinder, see file <cite>best_solution.tsv</cite>.</p></li>
<li><p><strong>all_systems.tsv</strong> - This file contains all possible candidate systems given the definitions -
without processing of the potential overlaps between candidate systems, under the form of a per-component, tabulated report file. It corresponds
to the tabulated version of the <cite>all_systems.txt</cite> file.</p></li>
</ul>
</div></blockquote>
<div class="section" id="all-systems-txt">
<span id="id1"></span><h4>all_systems.txt<a class="headerlink" href="#all-systems-txt" title="Permalink to this headline">¶</a></h4>
<p>The file starts with some comments:</p>
<blockquote>
<div><ul class="simple">
<li><p>the version of MacSyFinder used</p></li>
<li><p>the command line used to produce this file</p></li>
</ul>
</div></blockquote>
<p>Then for each replicon, the systems detected are listed along with their description:</p>
<blockquote>
<div><ul>
<li><p><strong>system_id</strong> - the unique identifier of a system</p></li>
<li><p><strong>model</strong> - the model assigned to this system</p></li>
<li><p><strong>replicon</strong> - the name of the replicon harbouring the system</p></li>
<li><p><strong>clusters</strong> - the clusters composition of this system</p>
<blockquote>
<div><ul>
<li><p>each clusters is a list of tuple</p></li>
<li><p>each tuple is composed of:</p>
<blockquote>
<div><ul class="simple">
<li><p>the name of the matching gene(s) in the replicon</p></li>
<li><p>the name of the corresponding gene profile(s)</p></li>
<li><p>the position of the corresponding sequence(s) along the replicon</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p><strong>occurrence</strong> - the average number of occurrences of each components of the system (as a potential proxy to estimate whether there’s the genetic potential for multiple systems in one)</p></li>
<li><p><strong>wholeness</strong> - the percentage of the model’s components that were found in this system</p></li>
<li><p><strong>loci nb</strong> - the number of different loci constituting this system</p></li>
<li><p><strong>score</strong> - the score of the system. See <a class="reference internal" href="functioning.html#combinatorial-exploration"><span class="std std-ref">here</span></a> for more details</p></li>
<li><p><strong>systems components</strong> - the number of occurrences of each model components
in parenthesis the name of the matching profile
in square brackets the name of other putative systems that would involve this gene</p></li>
</ul>
</div></blockquote>
<p>Here is an example of the <cite>all_systems.txt</cite> file:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyfinder 20200217.dev
# macsyfinder --sequence-db DATA_TEST/sequences.prt --db-type=gembase --models-dir data/models/ --models TFF-SF_final all -w 4
# Systems found:

system id = VICH001.B.00001.C001_MSH_1
model = TFF-SF_final/MSH
replicon = VICH001.B.00001.C001
clusters = [(&#39;VICH001.B.00001.C001_00406&#39;, &#39;MSH_mshI&#39;, 366), (&#39;VICH001.B.00001.C001_00407&#39;, &#39;MSH_mshJ&#39;, 367), (&#39;VICH001.B.00001.C001_00408&#39;, &#39;MSH_mshK&#39;, 368), (&#39;VICH001.B.00001.C001_00409&#39;, &#39;
MSH_mshL&#39;, 369), (&#39;VICH001.B.00001.C001_00410&#39;, &#39;MSH_mshM&#39;, 370), (&#39;VICH001.B.00001.C001_00411&#39;, &#39;MSH_mshN&#39;, 371), (&#39;VICH001.B.00001.C001_00412&#39;, &#39;MSH_mshE&#39;, 372), (&#39;VICH001.B.00001.C001_0041
3&#39;, &#39;MSH_mshG&#39;, 373), (&#39;VICH001.B.00001.C001_00414&#39;, &#39;MSH_mshF&#39;, 374), (&#39;VICH001.B.00001.C001_00415&#39;, &#39;MSH_mshB&#39;, 375), (&#39;VICH001.B.00001.C001_00416&#39;, &#39;MSH_mshA&#39;, 376), (&#39;VICH001.B.00001.C001
_00417&#39;, &#39;MSH_mshC&#39;, 377), (&#39;VICH001.B.00001.C001_00418&#39;, &#39;MSH_mshD&#39;, 378), (&#39;VICH001.B.00001.C001_00419&#39;, &#39;MSH_mshO&#39;, 379), (&#39;VICH001.B.00001.C001_00420&#39;, &#39;MSH_mshP&#39;, 380), (&#39;VICH001.B.00001
.C001_00421&#39;, &#39;MSH_mshQ&#39;, 381)]
occ = 1
wholeness = 0.941
loci nb = 1
score = 10.500

mandatory genes:
        - MSH_mshA: 1 (MSH_mshA)
        - MSH_mshE: 1 (MSH_mshE)
        - MSH_mshG: 1 (MSH_mshG)
        - MSH_mshL: 1 (MSH_mshL)
        - MSH_mshM: 1 (MSH_mshM)

accessory genes:
        - MSH_mshB: 1 (MSH_mshB)
        - MSH_mshC: 1 (MSH_mshC)
        - MSH_mshD: 1 (MSH_mshD)
        - MSH_mshF: 1 (MSH_mshF)
        - MSH_mshI: 1 (MSH_mshI)
        - MSH_mshI2: 0 ()
        - MSH_mshJ: 1 (MSH_mshJ)
        - MSH_mshK: 1 (MSH_mshK)
        - MSH_mshN: 1 (MSH_mshN)
        - MSH_mshO: 1 (MSH_mshO)
        - MSH_mshQ: 1 (MSH_mshQ)
        - MSH_mshP: 1 (MSH_mshP)

neutral genes:

============================================================
system id = VICH001.B.00001.C001_T4P_14
model = TFF-SF_final/T4P
replicon = VICH001.B.00001.C001
clusters = [(&#39;VICH001.B.00001.C001_00476&#39;, &#39;T4P_pilT&#39;, 427), (&#39;VICH001.B.00001.C001_00477&#39;, &#39;T4P_pilU&#39;, 428)], [(&#39;VICH001.B.00001.C001_00847&#39;, &#39;T4P_pilO&#39;, 778), (&#39;VICH001.B.00001.C001_00850&#39;,
 &#39;T4P_pilE&#39;, 781), (&#39;VICH001.B.00001.C001_00851&#39;, &#39;T4P_fimT&#39;, 782), (&#39;VICH001.B.00001.C001_00852&#39;, &#39;T4P_pilW&#39;, 783), (&#39;VICH001.B.00001.C001_00853&#39;, &#39;T4P_pilX&#39;, 784), (&#39;VICH001.B.00001.C001_00
854&#39;, &#39;T4P_pilV&#39;, 785)], [(&#39;VICH001.B.00001.C001_02305&#39;, &#39;T4P_pilA&#39;, 2202), (&#39;VICH001.B.00001.C001_02306&#39;, &#39;T4P_pilB&#39;, 2203), (&#39;VICH001.B.00001.C001_02307&#39;, &#39;T4P_pilC&#39;, 2204), (&#39;VICH001.B.000
01.C001_02308&#39;, &#39;T4P_pilD&#39;, 2205)], [(&#39;VICH001.B.00001.C001_02502&#39;, &#39;MSH_mshM&#39;, 2391), (&#39;VICH001.B.00001.C001_02505&#39;, &#39;T4P_pilQ&#39;, 2394), (&#39;VICH001.B.00001.C001_02506&#39;, &#39;T4P_pilP&#39;, 2395), (&#39;VI
CH001.B.00001.C001_02507&#39;, &#39;T4P_pilO&#39;, 2396), (&#39;VICH001.B.00001.C001_02508&#39;, &#39;T4P_pilN&#39;, 2397), (&#39;VICH001.B.00001.C001_02509&#39;, &#39;T4P_pilM&#39;, 2398)]
occ = 1
wholeness = 0.944
loci nb = 4
score = 12.000

mandatory genes:
        - T4P_pilE: 1 (T4P_pilE)
        - T4P_pilB: 1 (T4P_pilB)
        - T4P_pilC: 1 (T4P_pilC)
        - T4P_pilO: 2 (T4P_pilO, T4P_pilO)
        - T4P_pilQ: 1 (T4P_pilQ)
        - T4P_pilN: 1 (T4P_pilN)
        - T4P_pilT: 1 (T4P_pilT)
        - T4P_pilD: 1 (T4P_pilD [VICH001.B.00001.C001_T2SS_4])

accessory genes:
        - T4P_pilA: 1 (T4P_pilA)
        - T4P_pilV: 1 (T4P_pilV)
        - T4P_pilY: 0 ()
        - T4P_pilW: 1 (T4P_pilW)
        - T4P_pilX: 1 (T4P_pilX)
        - T4P_fimT: 1 (T4P_fimT)
        - T4P_pilM: 1 (T4P_pilM)
        - T4P_pilP: 1 (T4P_pilP)
        - T4P_pilU: 1 (T4P_pilU)
        - MSH_mshM: 1 (MSH_mshM)

neutral genes:
</pre></div>
</div>
</div>
<div class="section" id="all-systems-tsv">
<span id="id2"></span><h4>all_systems.tsv<a class="headerlink" href="#all-systems-tsv" title="Permalink to this headline">¶</a></h4>
<p>This corresponds to the tabulated version of the systems listed in <cite>all_systems.txt</cite>.
Each line corresponds to a “hit” that has been assigned to a detected system. It includes:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>replicon</strong> - the name of the replicon it belongs to</p></li>
<li><p><strong>hit_id</strong> - the unique identifier of the hit</p></li>
<li><p><strong>gene_name</strong> - the name of the component identified by the hit</p></li>
<li><p><strong>hit_pos</strong> - the position of the sequence in the replicon</p></li>
<li><p><strong>model_fqn</strong> - the model fully-qualified name</p></li>
<li><p><strong>sys_id</strong> - the unique identifier attributed to the detected system</p></li>
<li><p><strong>sys_loci</strong> - the number of loci</p></li>
<li><p><strong>sys_wholeness</strong> - the wholeness of the system</p></li>
<li><p><strong>sys_score</strong> - the system score</p></li>
<li><p><strong>sys_occ</strong> - the estimated number of system occurrences that could be potentially “filled” with this system’s occurrence, based on the average number of each component found. A proxy for the genetic potential ton encode several systems from the set of components found in this one occurrence.</p></li>
<li><p><strong>hit_gene_ref</strong> - the gene in the model whose this hit plays the role of</p></li>
<li><p><strong>hit_status</strong> - the status of the component in the assigned system’s definition</p></li>
<li><p><strong>hit_seq_len</strong> - the length of the protein sequence matched by this hit</p></li>
<li><p><strong>hit_i_eval</strong> - Hmmer statistics, the independent-evalue</p></li>
<li><p><strong>hit_score</strong> - Hmmer score</p></li>
<li><p><strong>hit_profile_cov</strong> - the percentage of the profile covered by the alignment with the sequence</p></li>
<li><p><strong>hit_seq_cov</strong> - the percentage of the sequence covered by the alignment with the profile</p></li>
<li><p><strong>hit_begin_match</strong> - the position in the sequence where the profile match begins</p></li>
<li><p><strong>hit_end_match</strong> - the position in the sequence where the profile match ends</p></li>
<li><p><strong>used_in</strong> - whether the hit could be used in another system’s occurrence</p></li>
</ul>
</div></blockquote>
<p>This file can be easily parsed using the Python <a class="reference external" href="https://pandas.pydata.org/">pandas</a> library.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">systems</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;path/to/systems.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>each system reported is separated from the others with a blank line to ease human reading.
These lines are ignored during the parsing with pandas.</p>
</div>
</div>
<div class="section" id="best-solution-tsv-and-all-best-solutions-tsv">
<h4>best_solution.tsv and all_best_solutions.tsv<a class="headerlink" href="#best-solution-tsv-and-all-best-solutions-tsv" title="Permalink to this headline">¶</a></h4>
<p>Since MacSyFinder 2.0, a combinatorial exploration of solutions using sets of systems found is performed. We call best solution, the combination of systems offering the highest score.</p>
<p>The <cite>best_solution.tsv</cite> and <cite>all_best_solutions.tsv</cite> files have the same structure as the file <cite>all_systems.tsv</cite>, except that there is an extra column <strong>sol_id</strong> which is a
solution identifier added to the file <cite>all_best_solutions.tsv</cite>. The systems that have the same “sol_id” belong to a same solution.</p>
<p>As the files have the same structure as <cite>all_systems.tsv</cite>, they can also be parsed with pandas as shown above.</p>
<p>For the description of the fields of <cite>best_solution.tsv</cite>, see <a class="reference internal" href="#all-systems-tsv"><span class="std std-ref">above</span></a> those of the <cite>all_systems.tsv</cite> file.</p>
<p>For the <cite>all_best_solutions.tsv</cite>, each line corresponds to a “hit” that has been assigned to a detected system. It includes:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>sol_id</strong> - the name of the solution it is part of</p></li>
<li><p><strong>replicon</strong> - the name of the replicon it belongs to</p></li>
<li><p><strong>hit_id</strong> - the unique identifier of the hit</p></li>
<li><p><strong>gene_name</strong> - the name of the component identified by the hit</p></li>
<li><p><strong>hit_pos</strong> - the position of the sequence in the replicon</p></li>
<li><p><strong>model_fqn</strong> - the model fully-qualified name</p></li>
<li><p><strong>sys_id</strong> - the unique identifier attributed to the detected system</p></li>
<li><p><strong>sys_loci</strong> - the number of loci</p></li>
<li><p><strong>sys_wholeness</strong> - the wholeness of the system</p></li>
<li><p><strong>sys_score</strong> - the system score</p></li>
<li><p><strong>sys_occ</strong> - the estimated number of system occurrences that could be potentially “filled” with this system’s occurrence, based on the average number of each component found. A proxy for the genetic potential ton encode several systems from the set of components found in this one occurrence.</p></li>
<li><p><strong>hit_gene_ref</strong> - the gene in the model whose this hit plays the role of</p></li>
<li><p><strong>hit_status</strong> - the status of the component in the assigned system’s definition</p></li>
<li><p><strong>hit_seq_len</strong> - the length of the protein sequence matched by this hit</p></li>
<li><p><strong>hit_i_eval</strong> - Hmmer statistics, the independent-evalue</p></li>
<li><p><strong>hit_score</strong> - Hmmer score</p></li>
<li><p><strong>hit_profile_cov</strong> - the percentage of the profile covered by the alignment with the sequence</p></li>
<li><p><strong>hit_seq_cov</strong> - the percentage of the sequence covered by the alignment with the profile</p></li>
<li><p><strong>hit_begin_match</strong> - the position in the sequence where the profile match begins</p></li>
<li><p><strong>hit_end_match</strong> - the position in the sequence where the profile match ends</p></li>
<li><p><strong>used_in</strong> - whether the hit could be used in another system’s occurrence</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you want to have a concise view of which systems have been found in your replicons and how many per replicon,
you can do it with a few lines of pandas on the <strong>best_solution.tsv</strong> file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">best_sol</span> <span class="o">=</span> <span class="s1">&#39;&lt;macsyfinder_results_dir&gt;/best_solution.tsv&#39;</span>

<span class="c1"># read data from best_solution file</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">best_sol</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>

<span class="c1"># remove useless columns</span>
<span class="n">selection</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span><span class="s1">&#39;replicon&#39;</span><span class="p">,</span> <span class="s1">&#39;sys_id&#39;</span><span class="p">,</span> <span class="s1">&#39;model_fqn&#39;</span><span class="p">]]</span>

<span class="c1"># keep only one row per replicon, sys_id</span>
<span class="n">dropped</span> <span class="o">=</span> <span class="n">selection</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">,</span> <span class="s1">&#39;sys_id&#39;</span><span class="p">])</span>

<span class="c1"># count for each replicon which models have been detected and their occurrences</span>
<span class="n">summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">dropped</span><span class="o">.</span><span class="n">replicon</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">dropped</span><span class="p">[</span><span class="s1">&#39;model_fqn&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>Below, an example of the result of these few lines of <em>pandas</em>:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>    model_fqn      TFF-SF/MSH  TFF-SF/T2SS  TFF-SF/T4P  TFF-SF/T4bP  TFF-SF/Tad
replicon
GCF_000005845           0            1           1            0           0
GCF_000006725           0            1           1            0           0
GCF_000006745           1            1           2            1           0
GCF_000006765           0            3           1            0           1
GCF_000006845           0            0           1            0           0
GCF_000006905           0            1           0            0           1
GCF_000006925           0            0           1            0           0
GCF_000006945           0            0           1            0           0
</pre></div>
</div>
<p>if you are not fluent in <cite>pandas</cite>, we provide you a tiny script <cite>msf_summary.py</cite> based on few lines above
to do the job</p>
<p><a class="reference download internal" download="" href="../_downloads/3774f623e8c0c5f9aa7fd1744abc32a8/msf_summary.py"><code class="xref download docutils literal notranslate"><span class="pre">msf_summary.py</span></code></a> .</p>
<p>Before to execute this script, you <strong>MUST</strong> install <a class="reference external" href="https://pandas.pydata.org/">pandas</a>.
If you run macsyfinder in a virtualenv do <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">pandas</span></code>
Then you can run the script</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">msf_summary</span><span class="o">.</span><span class="n">py</span> <span class="o">&lt;</span><span class="n">path_to_best_solution</span><span class="o">.</span><span class="n">tsv</span> <span class="ow">or</span> <span class="n">path_to_all_best_solutions</span><span class="o">.</span><span class="n">tsv</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>below an example of summary of a <cite>best_solution.tsv</cite> file</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>replicon	TFF-SF/MSH	TFF-SF/T2SS	TFF-SF/T4P	TFF-SF/T4bP	TFF-SF/Tad
GCF_000005845	0	1	1	0	0
GCF_000006725	0	1	1	0	0
GCF_000006745	1	1	2	1	0
GCF_000006765	0	3	1	0	1
GCF_000006845	0	0	1	0	0
GCF_000006905	0	1	0	0	1
GCF_000006925	0	0	1	0	0
GCF_000006945	0	0	1	0	0
</pre></div>
</div>
<p>and a summary of <cite>all_best_solutions.tsv</cite> correponding to the same <cite>macsyfinder</cite> run</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>sol_id	replicon	TFF-SF/MSH	TFF-SF/T2SS	TFF-SF/T4P	TFF-SF/T4bP	TFF-SF/Tad
1	GCF_000005845	0	1	1	0	0
2	GCF_000006725	0	1	1	0	0
3	GCF_000006725	0	1	1	0	0
4	GCF_000006745	1	1	2	1	0
5	GCF_000006745	1	1	2	1	0
6	GCF_000006745	1	1	1	1	0
7	GCF_000006765	0	3	1	0	1
8	GCF_000006845	0	0	1	0	0
9	GCF_000006905	0	1	0	0	1
10	GCF_000006925	0	0	1	0	0
11	GCF_000006945	0	0	1	0	0
</pre></div>
</div>
</div>
</div>
<div class="section" id="rejected-clusters-txt">
<span id="rejected-clusters-file"></span><h4>rejected_clusters.txt<a class="headerlink" href="#rejected-clusters-txt" title="Permalink to this headline">¶</a></h4>
<p>This file records all clusters or cluster combinations (if the “multi_loci” search mode is on) which have been discarded and the reason
why they were not selected as systems.</p>
<p>The header is composed of the MacSyFinder version and the command line used
followed by the description of the cluster(s). The list of the hits composing the cluster is presented
at the end of the cluster or clusters’ combination, followed by the reason why it has been discarded.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyfinder 20200511.dev
# /macsyfinder --sequence-db data/base/GCF_000006745.fasta --models TFF-SF all --models-dir data/models/ --db-type gembase -w 4
# Rejected clusters:

Cluster:
    - model: T4P
    - hits: (GCF_000005845_025680, T4P_pilW, 2568), (GCF_000005845_025690, T4P_fimT, 2569)
Cluster:
    - model: T4P
    - hits: (GCF_000005845_026930, T2SS_gspO, 2693)
Cluster:
    - model: T4P
    - hits: (GCF_000005845_030080, T2SS_gspO, 3008)
These clusters have been rejected because:
The quorum of mandatory genes required (4) is not reached: 1
The quorum of genes required (5) is not reached: 3
============================================================
Cluster:
    - model: Archaeal-T4P
    - hits: (GCF_000005845_019260, Archaeal-T4P_arCOG00589, 1926), (GCF_000005845_019310, Archaeal-T4P_arCOG02900, 1931)
These clusters have been rejected because:
The quorum of mandatory genes required (3) is not reached: 0
The quorum of genes required (3) is not reached: 2
============================================================
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="output-files-for-the-unordered-replicon-search-mode">
<span id="unordered-outputs"></span><h2>Output files for the “unordered replicon” search mode<a class="headerlink" href="#output-files-for-the-unordered-replicon-search-mode" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id4">
<h3>Systems detection results<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>As for ordered replicons, several output files are provided.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>all_possible_systems.txt</strong> - This file contains the description of candidate systems found.</p></li>
<li><p><strong>all_possible_systems.tsv</strong> - The same information as in <cite>all_possible_systems.txt</cite> but in the tabulated tsv format.</p></li>
<li><p><strong>uncomplete_systems.txt</strong> - This file contains occurrences for systems that did not complete models’ definitions and that were therefore not kept as candidate systems.</p></li>
</ul>
</div></blockquote>
<p>In this <cite>unordered</cite> search mode, there is no notion of order or distance of the components along the replicon. The clustering step
is skipped by MacSyFinder, and it is therefore “only” checked for each type of system being searched whether there is the genetic potential to fulfil its model definition.</p>
<div class="section" id="id5">
<h4>all_systems.txt<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h4>
<p>This file contains potential systems for unordered replicon in human readable format.</p>
<p>In this file, for each component of each searched system’s model, we report the number of hits found. For the description of the fields, see <a class="reference internal" href="#all-systems-txt"><span class="std std-ref">above</span></a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>In this mode the <cite>forbidden</cite> genes are reported here to the user. As we do not know if they co-localize (cluster) with the other genes they could
be present in the replicon, yet far away - or very close on the contrary - to the potential system.</p>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyfinder 20201028.dev
# macsyfinder --sequence-db tests/data/base/one_replicon.fasta --db-type unordered --models-dir tests/data/models -m TFF-SF T4P_single_locus
# Systems found:

This replicon contains genetic materials needed for system TFF-SF/T4P_single_locus


system id = Unordered_T4P_single_locus_1
model = TFF-SF/T4P_single_locus
replicon = Unordered
hits = [(&#39;GCF_000006845_000250&#39;, &#39;T4P_pilY&#39;, 25), (&#39;GCF_000006845_000700&#39;, &#39;T4P_pilY&#39;, 70), (&#39;GCF_000006845_001030&#39;, &#39;T4P_pilQ&#39;, 103), (&#39;GCF_000006845_001040&#39;, &#39;T4P_pilP&#39;, 104), (&#39;GCF_000006845_001050&#39;, &#39;T4P_pilO&#39;, 105), (&#39;GCF_000006845_001060&#39;, &#39;T4P_pilN&#39;, 106), (&#39;GCF_000006845_001070&#39;, &#39;T4P_pilM&#39;, 107), (&#39;GCF_000006845_003200&#39;, &#39;T4P_pilU&#39;, 320), (&#39;GCF_000006845_004190&#39;, &#39;T4P_fimT&#39;, 419), (&#39;GCF_000006845_004200&#39;, &#39;T4P_pilV&#39;, 420), (&#39;GCF_000006845_004210&#39;, &#39;T4P_pilW&#39;, 421), (&#39;GCF_000006845_004220&#39;, &#39;T4P_pilX&#39;, 422), (&#39;GCF_000006845_004230&#39;, &#39;T4P_pilA&#39;, 423), (&#39;GCF_000006845_010160&#39;, &#39;T4P_pilA&#39;, 1016), (&#39;GCF_000006845_012440&#39;, &#39;T4P_pilA&#39;, 1244), (&#39;GCF_000006845_014270&#39;, &#39;T4P_pilC&#39;, 1427), (&#39;GCF_000006845_014280&#39;, &#39;T4P_pilD&#39;, 1428), (&#39;GCF_000006845_014310&#39;, &#39;T4P_pilB&#39;, 1431), (&#39;GCF_000006845_016430&#39;, &#39;T4P_pilT&#39;, 1643), (&#39;GCF_000006845_016440&#39;, &#39;T4P_pilU&#39;, 1644)]
wholeness = 0.889

mandatory genes:
    - T4P_pilE: 0 ()
    - T4P_pilB: 1 (T4P_pilB)
    - T4P_pilC: 1 (T4P_pilC)
    - T4P_pilO: 1 (T4P_pilO)
    - T4P_pilQ: 1 (T4P_pilQ)
    - T4P_pilN: 1 (T4P_pilN)
    - T4P_pilT: 1 (T4P_pilT)
    - T4P_pilD: 1 (T4P_pilD)

accessory genes:
    - T4P_pilA: 3 (T4P_pilA, T4P_pilA, T4P_pilA)
    - T4P_pilV: 1 (T4P_pilV)
    - T4P_pilY: 2 (T4P_pilY, T4P_pilY)
    - T4P_pilW: 1 (T4P_pilW)
    - T4P_pilX: 1 (T4P_pilX)
    - T4P_fimT: 1 (T4P_fimT)
    - T4P_pilM: 1 (T4P_pilM)
    - T4P_pilP: 1 (T4P_pilP)
    - T4P_pilU: 2 (T4P_pilU, T4P_pilU)
    - MSH_mshM: 0 ()

neutral genes:

forbidden genes:

Use ordered replicon to have better prediction.
</pre></div>
</div>
</div>
<div class="section" id="all-systems-tsv-unordered">
<span id="id6"></span><h4>all_systems.tsv<a class="headerlink" href="#all-systems-tsv-unordered" title="Permalink to this headline">¶</a></h4>
<p>This file contains the same information as in <cite>all_systems.txt</cite> but in <cite>tsv</cite> format. For the description of the fields, see <a class="reference internal" href="#all-systems-tsv"><span class="std std-ref">above</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This file can be easily parsed with pandas:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">pot_systems</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;all_possible_systems.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyfinder 20201028.dev
# macsyfinder --sequence-db tests/data/base/one_replicon.fasta --db-type unordered --models-dir tests/data/models -m TFF-SF T4P_single_locus
# Likely Systems found:

replicon    hit_id  gene_name       hit_pos model_fqn       sys_id  sys_wholeness   hit_gene_ref    hit_status      hit_seq_len     hit_i_eval      hit_score       hit_profile_cov hit_seq_cov     hit_begin_match hit_end_match   used_in
Unordered   GCF_000006845_014310    T4P_pilB        1431    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilB        mandatory       558     3.8e-178        589.000 0.964   0.731   146     553
Unordered   GCF_000006845_014270    T4P_pilC        1427    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilC        mandatory       410     1.9e-131        434.800 0.997   0.817   72      406
Unordered   GCF_000006845_014280    T4P_pilD        1428    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilD        mandatory       286     2.8e-82 272.300 1.000   0.829   28      264
Unordered   GCF_000006845_001060    T4P_pilN        106     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilN        mandatory       199     2.3e-33 112.200 0.986   0.714   7       148
Unordered   GCF_000006845_001050    T4P_pilO        105     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilO        mandatory       215     2.9e-37 124.800 0.980   0.693   23      171
Unordered   GCF_000006845_001030    T4P_pilQ        103     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilQ        mandatory       723     1.9e-62 206.600 0.935   0.238   548     719
Unordered   GCF_000006845_016430    T4P_pilT        1643    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilT        mandatory       347     6.9e-167        551.400 0.997   0.983   2       342
Unordered   GCF_000006845_004190    T4P_fimT        419     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_fimT        accessory       221     2.7e-23 78.900  0.985   0.294   7       71
Unordered   GCF_000006845_004230    T4P_pilA        423     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilA        accessory       162     8.6e-20 67.800  0.744   0.389   9       71
Unordered   GCF_000006845_010160    T4P_pilA        1016    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilA        accessory       149     1.3e-15 54.300  0.821   0.430   5       68
Unordered   GCF_000006845_012440    T4P_pilA        1244    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilA        accessory       129     1.5e-19 67.000  0.859   0.519   6       72
Unordered   GCF_000006845_001070    T4P_pilM        107     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilM        accessory       371     3.3e-43 144.300 0.988   0.429   30      188
Unordered   GCF_000006845_001040    T4P_pilP        104     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilP        accessory       181     2.7e-34 115.600 1.000   0.735   13      145
Unordered   GCF_000006845_003200    T4P_pilU        320     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilU        accessory       376     2.2e-170        562.600 0.985   0.896   16      352
Unordered   GCF_000006845_016440    T4P_pilU        1644    TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilU        accessory       408     1.5e-127        421.800 0.994   0.833   40      379
Unordered   GCF_000006845_004200    T4P_pilV        420     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilV        accessory       203     9.6e-16 54.600  1.000   0.276   14      69
Unordered   GCF_000006845_004210    T4P_pilW        421     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilW        accessory       326     1.7e-10 38.000  0.517   0.190   17      78
Unordered   GCF_000006845_004220    T4P_pilX        422     TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilX        accessory       203     2.8e-18 62.600  0.983   0.286   17      74
Unordered   GCF_000006845_000250    T4P_pilY        25      TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilY        accessory       1006    2.2e-57 191.700 0.728   0.389   463     853
Unordered   GCF_000006845_000700    T4P_pilY        70      TFF-SF/T4P_single_locus Unordered_T4P_single_locus_1    0.889   T4P_pilY        accessory       1047    1.9e-57 191.900 0.721   0.362   516     894
</pre></div>
</div>
</div>
<div class="section" id="uncomplete-systems-txt">
<h4>uncomplete_systems.txt<a class="headerlink" href="#uncomplete-systems-txt" title="Permalink to this headline">¶</a></h4>
<p>This file is created when a search is performed in the <cite>unordered replicon</cite> mode.
This file list models that probably do not have not full systems in the replicon(s).
For each model, the reason why it is not fulfilled is reported,
followed by the model description and the components found.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyfinder 20201113.dev
# macsyfinder --sequence-db tests/data/base/one_replicon.fasta --db-type unordered --models-dir tests/data/models -m TFF-SF all
# Unlikely Systems found:

This replicon probably not contains a system TFF-SF/T2SS:
The quorum of mandatory genes required (4) is not reached: 1
The quorum of genes required (6) is not reached: 2

system id = Unordered_T2SS_3
model = TFF-SF/T2SS
replicon = Unordered
hits = [(&#39;GCF_000006845_002600&#39;, &#39;Tad_tadD&#39;, 260), (&#39;GCF_000006845_014280&#39;, &#39;T4P_pilD&#39;, 1428), (&#39;GCF_000006845_016430&#39;, &#39;T4P_pilT&#39;, 1643)]
wholeness = 0.143

mandatory genes:
        - T2SS_gspD: 0 ()
        - T2SS_gspE: 0 ()
        - T2SS_gspF: 0 ()
        - T2SS_gspG: 0 ()
        - T2SS_gspC: 0 ()
        - T2SS_gspO: 1 (T4P_pilD)

accessory genes:
        - T2SS_gspM: 0 ()
        - T2SS_gspH: 0 ()
        - T2SS_gspI: 0 ()
        - T2SS_gspJ: 0 ()
        - T2SS_gspK: 0 ()
        - T2SS_gspN: 0 ()
        - T2SS_gspL: 0 ()
        - Tad_tadD: 1 (Tad_tadD)

neutral genes:

forbidden genes:
        - T4P_pilT: 1 (T4P_pilT)

Use ordered replicon to have better prediction.

============================================================
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="logs-and-configuration-files">
<h2>Logs and configuration files<a class="headerlink" href="#logs-and-configuration-files" title="Permalink to this headline">¶</a></h2>
<p>Three specific output files are systematically built, whatever the search mode, to store information on MacSyFinder’s execution:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>macsyfinder.conf</strong> - contains the configuration information of the run. It is useful to recover all the parameters used for the run.</p></li>
<li><p><strong>macsyfinder.log</strong> - the log file, contains raw information on the run. Please send it to us with any <strong>bug report</strong>.</p></li>
</ul>
</div></blockquote>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="model_definition.html" class="btn btn-neutral float-right" title="Macromolecular models" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="gembase_convention.html" class="btn btn-neutral" title="Gembase format" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2021, &#39;Institut Pasteur (Paris), CNRS&#39;

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    

  

  <script type="text/javascript" src="../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>